tests/wallet/test_singleton_store.py

```diff
@@ -0,0 +1,279 @@
+        assert isinstance(latest, SingletonCoinRecord)
+        await db.set_spent(latest.coin_id(), uint32(20), uint64(time.time()))
+        last_record = await db.get_record_by_coin_id(latest.coin_id())
+        assert isinstance(last_record, SingletonCoinRecord)
+        assert last_record.spent_height == uint32(20)
+
+        # Try to set spent for non-existing record
+        await db.set_spent(record_2.coin_id(), uint32(20), uint64(time.time()))
+        res = await db.get_record_by_coin_id(record_2.coin_id())
+        assert res is None
+
+
+@pytest.mark.asyncio
+async def test_set_confirmed() -> None:
+    async with DBConnection(1) as wrapper:
+        db = await WalletSingletonStore.create(wrapper)
+        record_1 = get_new_singleton_record()
+        await db.add_singleton_record(record_1)
+        # Don't add the second record
+        record_2 = get_new_singleton_record()
+
+        # set record_1 confirmed (it has no parent)
+        await db.set_confirmed(record_1.coin_id(), uint32(10), uint64(time.time()))
+        res = await db.get_record_by_coin_id(record_1.coin.parent_coin_info)  # this is probably redundant
+        assert res is None
+
+        latest = await db.get_latest_singleton(record_1.singleton_id)
+        assert isinstance(latest, SingletonCoinRecord)
+        assert latest.confirmed
+
+        # set confirmed for non-existent record_2
+        await db.set_confirmed(record_2.coin_id(), uint32(10), uint64(time.time()))
+        res = await db.get_record_by_coin_id(record_2.coin_id())
+        assert res is None
+
+        # add record_2 and set confirmed
+        await db.add_singleton_record(record_2)
+        await db.set_confirmed(record_2.coin_id(), uint32(20), uint64(time.time()))
+
+        # get next record and add it
+        next_record_2 = get_next_singleton_record(record_2)
+        await db.add_singleton_record(next_record_2)
+        await db.set_confirmed(next_record_2.coin_id(), uint32(30), uint64(time.time()))
+
+        # check parent has been set to spent
+        parent_record = await db.get_record_by_coin_id(record_2.coin_id())
+        assert isinstance(parent_record, SingletonCoinRecord)
+        assert parent_record.spent_height == uint32(30)
+
+
+@pytest.mark.asyncio
+async def test_custom_data() -> None:
+    async with DBConnection(1, foreign_keys=True) as wrapper:
+        db = await WalletSingletonStore.create(wrapper)
+        custom_data = {"field_1": 100, "field_2": b"some text"}
+        record = get_new_singleton_record(custom_data)
+        await db.add_singleton_record(record)
+        res = await db.get_records_by_singleton_id(record.singleton_id)
+        assert res[0].custom_data == custom_data
+
+        # delete the singleton and check it's removed from custom data table
+        await db.delete_records_by_singleton_id(record.singleton_id)
+        custom_res = await db.get_custom_data_by_coin_id(record.coin_id())
```

Comment:
the only test of `get_custom_data_by_coin_id()` is that it can return `None`. More cases should be covered, especially a coin that exists, but also a coin that never existed.